archive/imo/imo2021_q1.lean

```diff
@@ -151,69 +151,158 @@ begin
+        rw [hxc, hya] },
+      { convert h₃,
+        rw [hxc, hyb, add_comm] },
+      { rw ← hyc at hxc,
+        contradiction }}},
+  { intros d hd,
+    simp only [finset.mem_insert, finset.mem_singleton] at hd,
+    split,
+    { rcases hd with (hd | hd | hd);
+      { linarith }},
+    { rcases hd with (hd | hd | hd);
+      { linarith }}},
+end
+
+-- If for n ∈ ℕ, and finite sets B, X, Y such that B ⊆ X ∪ Y and B has cardinality at least
+-- 2 * n + 1, then there exists a subset C ⊆ A of cardinality at least n + 1, which is entirely
+-- contained in X or Y. For our result, we will apply this lemma with n = 1 and any X such that
+-- X ⊆ [n, 2n] and Y = [n, 2n] \ X, while we'll let B ⊆ [n, 2n] be the subset such that each
+-- pairwise unequal pair of B sums to a perfect square.
+lemma finset_subset_or_complement_cardinality {α : Type*} [decidable_eq α] (B X Y : finset α)
+(h : B ⊆ X ∪ Y) {n : ℕ} (hB : 2 * n + 1 ≤ B.card) :
+  (∃ C, C ⊆ B ∧ C ⊆ X ∧ n + 1 ≤ C.card) ∨ (∃ C, C ⊆ B ∧ C ⊆ Y ∧ n + 1 ≤ C.card) :=
+begin
+  by_contra h,
+  push_neg at h,
+  cases h with h₁ h₂,
+  have h₃: B \ (Y \ X) ⊆ B := finset.sdiff_subset B (Y \ X),
+  have h₄: B \ (Y \ X) ⊆ X,
+  { have : B \ (Y \ X) ⊆ (X ∪ Y) \ (Y \ X) := finset.sdiff_subset_sdiff h (finset.subset.refl (Y \ X)),
+    apply finset.subset.trans this,
+    intros i hi,
+    simp only [not_and, not_not, finset.mem_sdiff, finset.mem_union] at hi,
+    cases hi with hi₁ hi₂,
+    cases hi₁,
+    { exact hi₁ },
+    { exact hi₂ hi₁ }},
+  have h₅ : B \ (X \ Y) ⊆ B := finset.sdiff_subset B (X \ Y),
+  have h₆: B \ (X \ Y) ⊆ Y,
+  { have : B \ (X \ Y) ⊆ (X ∪ Y) \ (X \ Y) := finset.sdiff_subset_sdiff h (finset.subset.refl (X \ Y)),
+    apply finset.subset.trans this,
+    intros i hi,
+    simp only [not_and, not_not, finset.mem_sdiff, finset.mem_union] at hi,
+    cases hi with hi₁ hi₂,
+    cases hi₁,
+    { exact hi₂ hi₁ },
+    { exact hi₁ }},
+  specialize h₁ (B \ (Y \ X)) h₃ h₄,
+  specialize h₂ (B \ (X \ Y)) h₅ h₆,
+  have h₇ : B ⊆ B \ (Y \ X) ∪ B \ (X \ Y),
+  { intros i hi,
+    simp only [not_and, not_not, finset.mem_sdiff, finset.mem_union],
+    have h₈ : i ∈ X ∨ i ∈ Y := finset.mem_union.mp (h hi),
+    cases h₈,
+    { left,
+      refine ⟨hi,_⟩,
+      { intro,
+        exact h₈ } },
+    { right,
+      refine ⟨hi,_⟩,
+      { intro,
+        exact h₈ }}},
+  replace h₇ := finset.card_le_of_subset h₇,
+  have h₉ : B.card ≤  2 * n,
+  { apply le_trans h₇ _,
+    apply le_trans ((B \ (Y \ X)).card_union_le (B \ (X \ Y))) _,
+    linarith },
+  linarith,
 end
 
+lemma finset_two_le_card_exists_pair_neq (C : finset ℕ) (hC : 2 ≤ C.card) :
+  ∃ (a b ∈ C), a ≠ b :=
+begin
+  by_contra hab,
+  push_neg at hab,
+  rw ← finset.card_le_one_iff at hab,
+  linarith,
+end
+
+
 theorem IMO_2021_Q1 : ∀ (n : ℕ), 100 ≤ n → ∀ (A ⊆ finset.Ico n (2 * n + 1)),
   (∃ (a b ∈ A), a ≠ b ∧ ∃ (k : ℕ), a + b = k * k) ∨
   (∃ (a b ∈ finset.Ico n (2 * n + 1) \ A), a ≠ b ∧ ∃ (k : ℕ), a + b = k * k) :=
```

Comment:
Again, I would advise you not to use `Ico` for membership alone. And you can merge those two `∃` like so:
```suggestion
  ∃ a b k, n ≤ a ∧ a ≤ 2 * n ∧ n ≤ b ≤ b ≤ 2 * n ∧ a ∉ A ∧ b ∉ A ∧ a ≠ b ∧ a + b = k * k :=
```

Reply:
I'd like to slightly push back on this one. Since this is the final statement of the problem, isn't it much better to have the statement as close as possible to the one in real life? Personally, I believe the present version is more ''real-life'' like.

Reply:
Sure, in this case it's a matter of taste.